nnet_utils/nnet_activation.h

```diff
@@ -137,69 +141,93 @@ inline float exp_fcn_float(float input) {
         //std::cout << "Lookup table In Value: " << in_val << " Result: " << real_val << std::endl;
         table_out[ii] = real_val;
     }
 }
 
+template<typename CONFIG_T, int N_TABLE>
+void init_invert_table(typename CONFIG_T::table_t table_out[N_TABLE])
+{
+    // Inversion function:
+    //   result = 1/x
+    for (int ii = 0; ii < N_TABLE; ii++) {
+      // First, convert from table index to X-value (signed 8-bit, range 0 to +64)
+	float in_val = 64.0*ii/float(N_TABLE);
+        // Next, compute lookup table function
+	if (in_val > 0.0) table_out[ii] = 1.0/in_val;
+	else table_out[ii] = 0.0;
+    }    
+}
+
 template<class data_T, class res_T, typename CONFIG_T>
 void  softmax(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
 {
     // Initialize the lookup table
-    res_T exp_table[CONFIG_T::table_size];
-    init_exp_table<res_T, CONFIG_T::table_size>(exp_table);
+    typename CONFIG_T::table_t exp_table[CONFIG_T::table_size];
+    init_exp_table<CONFIG_T, CONFIG_T::table_size>(exp_table);
+
+    typename CONFIG_T::table_t invert_table[CONFIG_T::table_size];
+    init_invert_table<CONFIG_T, CONFIG_T::table_size>(invert_table);
 
     if (CONFIG_T::io_type == io_parallel){
         // Note: This is going to be a resource hog to run with pipeline, but hey, whatever
         #pragma HLS PIPELINE
     }
 
     // Index into the lookup table based on data for exponentials
-    res_T exp_res[CONFIG_T::n_in];//same precision as rest?
-    res_T exp_res_sum=0;
-    data_T datareg;
+    typename CONFIG_T::table_t exp_res[CONFIG_T::n_in];// different, independent, fixed point precision
+    typename CONFIG_T::table_t exp_diff_res[CONFIG_T::n_in][CONFIG_T::n_in];// different, independent, fixed point precision
     int data_round;
     int index;
     for (int ii=0; ii<CONFIG_T::n_in; ii++) {
-        // #pragma HLS UNROLL
-        data_round = data[ii]*CONFIG_T::table_size/16;
-        index = data_round + 8*CONFIG_T::table_size/16;
-        if (index < 0)   index = 0;
-        if (index > CONFIG_T::table_size-1) index = CONFIG_T::table_size-1;
-        exp_res[ii] = exp_table[index];
-        exp_res_sum += exp_table[index];
+      exp_res[ii] = 0;
+      for (int jj=0; jj<CONFIG_T::n_in; jj++) {
+	if (ii==jj) exp_diff_res[ii][jj] = 1;
+	else {
+	  data_round = (data[ii]-data[jj])*CONFIG_T::table_size/16;
+	  index = data_round + 8*CONFIG_T::table_size/16;
+	  if (index < 0)   index = 0;
+	  if (index > CONFIG_T::table_size-1) index = CONFIG_T::table_size-1;
+	  exp_diff_res[ii][jj] = exp_table[index];
+	}
+	exp_res[ii] += exp_diff_res[ii][jj];
+      }
     }
 
-    //Second loop to divide by sum of exponentials
+    //Second loop to invert
     for (int ii=0; ii<CONFIG_T::n_in; ii++) {
-      // #pragma HLS UNROLL
-      res[ii] = exp_res[ii]/exp_res_sum; //Note division used here!
+      int exp_res_index = exp_res[ii]*CONFIG_T::table_size/64;
+      if (exp_res_index < 0)   exp_res_index = 0;
```

Comment:
It seems like here there is some sign flip happening on the exp_res due to the precision of the table which I am setting now as the precision of the rest. We know already about this problem. In fact, I plan to redo the scans by setting the precision of the table to something reasonable like <18,8>.

Reply:
@jngadiub it sounds like you figured it out.  The only other thought is that `table_size` could be playing a role, in which case increasing it would effectively increase precision of the outputs.  If changing table_t to <18,8> doesn't get you all the there, I'd try that.